base/operators.jl

```diff
@@ -157,6 +157,28 @@ isless(x::AbstractFloat, y::AbstractFloat) = (!isnan(x) & (isnan(y) | signless(x
 isless(x::Real,          y::AbstractFloat) = (!isnan(x) & (isnan(y) | signless(x, y))) | (x < y)
 isless(x::AbstractFloat, y::Real         ) = (!isnan(x) & (isnan(y) | signless(x, y))) | (x < y)
 
+"""
+    isgreater(x, y)
+
+Test whether `x` is greater than `y`, according to a fixed total order.
+`isgreater` is defined in terms of `isless`, but is not the opposite of that function.
+
+`isless` defines a fixed total order that ascends with unorderable values (such as `NaN`) and
+[`missing`](@ref) ordered last (biggest).
+`isgreater` defines a fixed total order that descends with unorderable values
+and `missing` ordered last (smallest).
+
+Values that are normally unordered, such as `NaN`,
+are ordered after regular values.
+[`missing`](@ref) values are ordered last.
+
+# Implementation
+Types should usually not implement this function. Instead, implement `isless`.
+"""
+isgreater(a, b) = _is_unorderable(a) || _is_unorderable(b) ? isless(a, b) : isless(b, a)
+_is_unorderable(x) = !isa(x == x, Bool) || x != x
```

Comment:
Maybe it should be called `_irreflexive` or `_anti_reflexive`?

Also, `==` and `!=` can be expensive for containers like vectors and strings. So how about

```julia
neq = x != x
return !isa(neq, Bool) || neq
```

?

Also, is it possible to prefer `missing` over `NaN` by factoring out irreflexivity check like this?  The reason why I did it with a multi-line definition in https://github.com/JuliaLang/julia/issues/27613#issuecomment-601502228 was that I thought it was impossible. (Though you can in principle define `isgreater` on `Missing` here.)

Reply:
I normally just trust constant propagation to sort that kind of thing out, but I could manually do it if it's sure to make a difference.

Reply:
For things like arrays, I don't think the compiler can do anything clever (imagine a worst-case scenario like `getindex` with `println` in it).

Reply:
Adjusted roughly as suggested. I'm not persuaded by the println usecase, but I guess getindex could involve moving data from GPU to CPU or over the network or whatever.

Reply:
It's very easy to check that running `!=` and `==` is twice as slow as running `==` once, even for plain array:

```julia
julia> f(x) = !isa(x == x, Bool) || x != x
f (generic function with 1 method)

julia> function g(x)
           neq = x != x
           return !isa(neq, Bool) || neq
       end
g (generic function with 1 method)

julia> x = rand(1000);

julia> @btime f($x)
  2.053 μs (0 allocations: 0 bytes)
false

julia> @btime g($x)
  1.028 μs (0 allocations: 0 bytes)
false
```

Reply:
I already made the change on the basis that it can't hurt and I hadn't benchmarked it yet :)